sdk/storagev2_internal.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	return bcDrives, nil
+}
+
+type driveV2DTO struct {
+	Drive                          string    `json:"replicator"`
+	LastApprovedDataModificationId *hashDto  `json:"lastApprovedDataModificationId"`
+	DataModificationIdIsValid      bool      `json:"dataModificationIdIsValid"`
+	InitialDownloadWork            uint64DTO `json:"initialDownloadWork"`
+}
+
+type driveV2DTOs []*driveV2DTO
+
+func (ref *driveV2DTOs) toStruct(networkType NetworkType) (map[string]*DriveInfo, error) {
+	var (
+		dtos      = *ref
+		driveInfo = make(map[string]*DriveInfo)
+	)
+
+	for i, dto := range dtos {
+		drive, err := NewAccountFromPublicKey(dto.Drive, networkType)
+		if err != nil {
+			return nil, err
+		}
+
+		lastApprovedDataModificationId, err := dto.LastApprovedDataModificationId.Hash()
+		if err != nil {
+			return nil, err
+		}
+
+		info := DriveInfo{
+			LastApprovedDataModificationId: lastApprovedDataModificationId,
+			DataModificationIdIsValid:      dto.DataModificationIdIsValid,
+			InitialDownloadWork:            dto.InitialDownloadWork.toUint64(),
+			Index:                          i,
+		}
+
+		driveInfo[drive.PublicKey] = &info
+	}
+
+	return driveInfo, nil
+}
+
+type replicatorV2DTO struct {
+	Replicator struct {
+		ReplicatorKey string      `json:"multisig"`
+		Version       int32       `json:"version"`
+		Capacity      uint64DTO   `json:"capacity"`
+		BLSKey        string      `json:"blsKey"`
+		Drives        driveV2DTOs `json:"drives"`
+	}
+}
+
+func (ref *replicatorV2DTO) toStruct(networkType NetworkType) (*Replicator, error) {
+	replicator := Replicator{}
+
+	replicatorAccount, err := NewAccountFromPublicKey(ref.Replicator.ReplicatorKey, networkType)
+	if err != nil {
+		return nil, err
+	}
+
+	replicator.ReplicatorKey = replicatorAccount
+	replicator.Version = ref.Replicator.Version
+	replicator.Capacity = ref.Replicator.Capacity.toStruct()
+	replicator.BLSKey = BLSPublicKey(ref.Replicator.BLSKey)
+
+	drives, err := ref.Replicator.Drives.toStruct(networkType)
+	if err != nil {
+		return nil, fmt.Errorf("sdk.replicatorV2DTO.toStruct Replicator.Drives.toStruct: %v", err)
+	}
+
+	replicator.Drives = drives
+
+	return &replicator, nil
+}
+
+type bcDrivesPageDTO struct {
+	BcDrives []bcDriveDTO `json:"data"`
+
+	Pagination struct {
+		TotalEntries uint64 `json:"totalEntries"`
+		PageNumber   uint64 `json:"pageNumber"`
+		PageSize     uint64 `json:"pageSize"`
+		TotalPages   uint64 `json:"totalPages"`
+	} `json:"pagination"`
+}
+
+func (t *bcDrivesPageDTO) toStruct(networkType NetworkType) (*BcDrivesPage, error) {
+	page := &BcDrivesPage{
+		BcDrives: make([]*BcDrive, len(t.BcDrives)),
+		Pagination: Pagination{
+			TotalEntries: t.Pagination.TotalEntries,
+			PageNumber:   t.Pagination.PageNumber,
+			PageSize:     t.Pagination.PageSize,
+			TotalPages:   t.Pagination.TotalPages,
+		},
+	}
+
+	errs := make([]error, len(t.BcDrives))
+	for i, t := range t.BcDrives {
+		currDr, currErr := t.toStruct(networkType)
```

Comment:
It's more efficient to see the errors for each drive entry in one go because each drive entry may or may not have an error. Instead of finding out if a particular drive entry has an error in the next test run.

Reply:
but below if there are errors will be returned only the first one. Even more, below can be returned not `nil` result and not `nil` error and it's a bit confusing

Reply:
I've changed it to return the error right away at https://github.com/proximax-storage/go-xpx-chain-sdk/pull/105/commits/a0b74d8e3cba5f9c5e9c0d7e45d15e1a4613d178

Reply:
Changes in another func :D If occurs an error then `return nil, err`

Reply:
Do you mean to change `return page, err` to `return nil, err`?
```
currDr, err := t.toStruct(networkType)
page.Replicators[i] = currDr
if err != nil {
	return nil, err
}
```

Reply:
Not really. I propose to replace this
```go
errs := make([]error, len(t.BcDrives))
	for i, t := range t.BcDrives {
		currDr, currErr := t.toStruct(networkType)
		page.BcDrives[i], errs[i] = currDr, currErr
	}

	for _, err := range errs {
		if err != nil {
			return page, err
		}
	}
```
to this
```go
       var err error
	for i, t := range t.BcDrives {
		page.BcDrives[i], err = t.toStruct(networkType)
		if err != nil {
				return nil, err
		}
	}
```

Reply:
and in other similar places do the same

Reply:
I've updated it in 0d4b537